forecast-archive/forecast_archive/app.py

```diff
@@ -0,0 +1,127 @@
+"""Delete the forecasts made before a certain date.
+
+This process can be batched and optionally we can wait between each batch to not overload the
+database with work.
+"""
+
+import contextlib
+import datetime as dt
+import logging
+import os
+import time
+import uuid
+
+import click
+import sqlalchemy as sa
+from pvsite_datamodel.sqlmodels import ForecastSQL, ForecastValueSQL
+from sqlalchemy.orm import Session, sessionmaker
+
+_log = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def _profile(msg: str):
+    _log.debug(msg)
+    t0 = time.time()
+    yield
+    t1 = time.time()
+    _log.debug(f"Done in {t1 - t0:.3f}s")
+
+
+def _get_forecasts(session: Session, max_date: dt.datetime, limit: int) -> list[uuid.UUID]:
+    """Get the `limit` older forecasts that are before `max_date`."""
+    stmt = (
+        sa.select(ForecastSQL.forecast_uuid)
+        .where(ForecastSQL.timestamp_utc < max_date)
+        .order_by(ForecastSQL.timestamp_utc)
+        .limit(limit)
+    )
+
+    return session.scalars(stmt).all()
+
+
+def _delete_forecasts_and_values(session: Session, forecast_uuids: list[uuid.UUID]) -> None:
+    """Delete the forecasts and their corresponding values."""
+    with _profile(f"Deleting forecast values for {len(forecast_uuids)} forecasts."):
+        stmt = sa.delete(ForecastValueSQL).where(ForecastValueSQL.forecast_uuid.in_(forecast_uuids))
+        session.execute(stmt)
+
+    with _profile(f"Deleting {len(forecast_uuids)} forecasts"):
+        stmt = sa.delete(ForecastSQL).where(ForecastSQL.forecast_uuid.in_(forecast_uuids))
+        session.execute(stmt)
+
+
+@click.command()
+@click.option(
+    "--date", required=True, help="Datetime (UTC) before which to delete, format: YYYY-MM-DD HH:mm"
+)
+@click.option(
+    "--batch-size",
+    default=10,
+    help="Number of forecasts to delete in one batch."
+    " (Note that this means orders of magnitude more Forecast *Values*).",
+    show_default=True,
+)
+@click.option(
+    "--sleep",
+    type=float,
+    default=0,
+    help="How much time to wait between batches in seconds",
+    show_default=True,
+)
+@click.option(
+    "--log-level",
+    default="info",
+    show_default=True,
+)
+@click.option(
+    "--do-delete", is_flag=True, help="Actually delete the rows. By default we only do a dry run."
+)
+def main(date: dt.datetime, batch_size: int, sleep: int, do_delete: bool, log_level: str):
+    logging.basicConfig(
+        level=getattr(logging, log_level.upper()),
+        format="[%(asctime)s] {%(pathname)s:%(lineno)d} %(levelname)s - %(message)s",
+    )
+
+    db_url = os.environ["DB_URL"]
+    engine = sa.create_engine(db_url)
+    Session = sessionmaker(engine, future=True)
+    max_date = dt.datetime.strptime(date, "%Y-%m-%d %H:%M")
+
+    num_forecast_deleted = 0
+
+    while True:
+        with Session.begin() as session:
+            forecast_uuids = _get_forecasts(
```

Comment:
I thought we could do a just one query, something like this - https://github.com/openclimatefix/nowcasting_datamodel/blob/f7a156fc2d92dbd4e9d76177678e759127e5a078/nowcasting_datamodel/save/update.py#L253

You can probably put a limit on that, if you want to do it in batches

Reply:
Ah, ignore this, then you would be able to save it to s3 for example. So stick with you method

Reply:
I can't just do one query because I need to delete from 2 tables.

We could do it in one if we properly setup the `CADASDE` effects in the database definition (so that deleting a `forecast` would delete the `forecast_values`).

Your point on saving though is making me think: should we instead write the data to S3 *when we do the forecast*, instead of when we delete old forecasts? This could potentially be simpler and we would always have up-to-date data in S3.